assignment5/unknownLanguage.py

```diff
@@ -0,0 +1,76 @@
+def unknownLanguage(dictionary):
+    """Given a dictionary of all words in an unknown/invented language, find the
+    alphabet of that language.
+
+    Args:
+        dictionary: a list of words in lexicographic order.
+
+    Returns:
+        the alphabet, an ordered list of characters.
+    """
+    def dfs(v, visited):
+        """
+        Depth-first search with cycle search.
+        """
+        visited[v] = 1
+        for u in adjList[v]:
+            if visited[u] == 0:
+                dfs(u, visited)
+            elif visited[u] == 1:
+                # in that case, we have cycles in our graph =>
+                # it can't be topological sorted => it's not an alphabet
+                raise ValueError("dictionary is not in lexicographic order")
+        visited[v] = 2
+        alphabet.append(v)
+
+    def topologicalSort():
+        """
+        An ordinary topological sort based on dfs.
+        """
+        visited = dict()
+        for v in adjList:
+            visited[v] = 0
+        for v in adjList:
+            if not visited[v]:
+                dfs(v, visited)
+        alphabet.reverse()
+
+
+    if not dictionary:
+        return []
+
+    # define adjacency list where nodes are letters, not nums
+    adjList = dict()
+    for word in dictionary:
+        for let in word:
+            if let not in adjList:
+                adjList[let] = set()
+
+    maxLen = max([len(word) for word in dictionary])
+
+    # bruteforce all possible prefixes by they length
+    for prefixLen in range(maxLen):
```

Comment:
Optimization:

This code is reading the dictionary multiples times. In fact as many times as the length of the longest word. In a dictionary with `N` words of length at most `k`, this is `O(kN)` (assuming look up into the dict and adding to the set are 'O(1)', which is reasonable).

However, given two consecutive words in the directory there is only one possible common prefix. So we can actually scan the directory only once instead.

If you think of:
```
all
almond
azure
```
the first two words have the prefix `al`, the second and third have the prefix `a`.
So if `j` is the length of the longest prefix, then the result is `O(jN)` where `j` is strictly smaller than `k` (assuming no repeated words) and in practice will actually be much better because many words will have a much shorter common prefix.

Try to turn the code around so that you look at the dictionary only once.